src/fparser/two/Fortran2003.py

```diff
@@ -6935,51 +6942,155 @@ def tostr(self):
+         the second item in the io-control-spec-list and the first item shall
+         be io-unit.
+    C919 If io-unit is not a file-unit-number, the io-control-spec-list shall
+         not contain a REC= specifier or a POS= specifier.
+    C920 If the REC= specifier appears, an END= specifier shall not appear, a
+         namelist-groupname shall not appear, and the format, if any, shall not
+         be an asterisk.
+    C921 An ADVANCE= specifier may appear only in a formatted sequential or
+         stream input/output statement with explicit format specification
+         (10.1) whose control information list does not contain an
+         internal-file-variable as the io-unit.
+    C922 If an EOR= specifier appears, an ADVANCE= specifier also shall appear.
+    C923 If a SIZE= specifier appears, an ADVANCE= specifier also shall appear.
+    C924 The scalar-char-initialization-expr in an ASYNCHRONOUS= specifier
+         shall be of type default character and shall have the value YES or NO.
+    C925 An ASYNCHRONOUS= specifier with a value YES shall not appear unless
+         io-unit is a file-unit-number.
+    C926 If an ID= specifier appears, an ASYNCHRONOUS= specifier with the value
+         YES shall also appear.
+    C927 If a POS= specifier appears, the io-control-spec-list shall not
+         contain a REC= specifier.
+    C928 If a DECIMAL=, BLANK=, PAD=, SIGN=, or ROUND= specifier appears, a
+         format or namelist-group-name shall also appear.
+    C929 If a DELIM= specifier appears, either format shall be an asterisk or
+         namelist-group-name shall appear.
+
+    TODO #267. Of these constraints, only C910 & C916-918 are currently
+    enforced.
+
+    '''
     subclass_names = []
-    use_names = ['Io_Control_Spec']
+    use_names = ['Io_Control_Spec', 'Namelist_Group_Name', 'Format']
 
     @staticmethod
     def match(string):
+        '''
+        Attempts to match the supplied string with a list of Io_Control_Spec
+        items. We have to override the base implementation because the first
+        two items in the list have specific meanings if they are not explictly
+        named: the first must be the unit number and the second may be either
+        a format specifier *or* a namelist-group-name.
+
+        :param str string: the string that is checked for a match.
+
+        :returns: a tuple of Io_Control_Spec objects if the match is \
+                  successful, None otherwise.
+        :rtype: tuple of :py:class:`fparser.two.Fortran2003.Io_Control_Spec` \
+                objects or NoneType
+
+        '''
         line, repmap = string_replace_map(string)
         splitted = line.split(',')
         lst = []
-        unit_is_positional = False
+        # If the first entry in the list is not named then it must be a
+        # unit number (C910)
+        spec = splitted.pop(0).strip()
+        spec = repmap(spec)
+        try:
+            io_spec = Io_Control_Spec_Unit(spec)
```

Comment:
I think that the robust way to do this is to sort out all un-named entries here and then limit `Io_Control_Spec` to deal with named entries. I think the way to do this is to check directly for a match with `io-unit` for the first argument (i.e. not use `Io_Control_Spec_Unit`) and then if this matches, check directly for a match with either `format` or `namelist-group-name` with the second argument. All other arguments (including non-matching second or first arguments) can then use `Io_Control_Spec` coded simply for named entries.

Reply:
Yes, I wanted to do it that way but hit a problem because of the way matching is mixed up with object creation. I need `Io_Control_Spec_List` to contain `Io_Control_Spec` objects but if I match directly on `io-unit` then I don't have such an object. Have you got any ideas as to how to get round this?

Reply:
Ah I see. How about matching on `Io_Unit` to determine if it is a 'positional io unit', then if it is a successful match creating an `Io_Control_Spec` object? Something like ...

```
if Io_Unit.match(string):
    positional_io_unit = True
    obj = Io_Control_Spec(string)
```

Reply:
I'm getting there now - things are nicer with this approach. However, I've discovered that
'FMT=\'(5X,"q_mesh =",1F12.8)\'' produces a match with `expr` and that results in me getting a match for `Format` when I shouldn't (because `Format` shouldn't match a named format). Looking at the standard, I can't see how 'lhs=rhs' is getting matched as an `expr` - must be another bug.

Reply:
Yes, this approach does rely on no false positives from the more specific rules. As you say, lhs=rhs should not match as a valid expression so is probably some loose matching somewhere. The worry is, of course, that there is a pattern that is valid for some reason. I could take a look if you want to create a separate issue?

Reply:
That's very kind of you - I've created #283.